appcelerator/titanium_mobile
# refactor(android): media/file APIs to use scoped storage

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-28183
- https://jira.appcelerator.org/browse/TIMOB-28059
- https://jira.appcelerator.org/browse/TIMOB-28146
- https://jira.appcelerator.org/browse/TIMOB-27481
- https://jira.appcelerator.org/browse/TIMOB-27201
- https://jira.appcelerator.org/browse/TIMOB-26602

**Summary:**
- [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059) Modified `Ti.Media.showCamera()` and `saveToPhotoGallery()` to use scoped storage.
  * No longer requires external storage permission on Android 10 and higher.
- Camera Overlay Changes:
  * [TIMOB-27481](https://jira.appcelerator.org/browse/TIMOB-27481) Fixed bug where navigating back would may wrongly exit out of the app.
  * [TIMOB-26602](https://jira.appcelerator.org/browse/TIMOB-26602) Fixed bug where `Ti.Media.takePicture()` may use the wrong file extension.
- Fixed `Ti.Media.VideoPlayer` crash which can occur if error callback is invoked during playback.
- [TIMOB-28183](https://jira.appcelerator.org/browse/TIMOB-28183) Added media methods:
  * `Ti.Media.hasPhotoGalleryPermissions()`
  * `Ti.Media.requestPhotoGalleryPermissions()`
- [TIMOB-27201](https://jira.appcelerator.org/browse/TIMOB-27201) Updated `Ti.Filesystem.File` APIs when wrapping "content://" URL.
  * Added: `createdAt()`, `deleteFile()`, `exists()`, `extension()`, `modifiedAt()`, `read()`, `readonly`, `write()`, `writable`
  * Missing: `createDirectory()`, `open()`, `readLine()`
- Optimized `Ti.Filesystem.File` object's `move()` method when performed on same volume.
- Updated `Ti.Filesystem.File` object's `rename()` to support absolute path in same folder.
- Renamed `TitaniumBlob.java` to `TiContentFile.java`.
  * New name better indicates it supports "content://" URLs.
- Removed unused "ti.android.enablecoverage" related code. _(Was never supported.)_

---
**Camera Photo Test:**
_(Run this test on Android 11 and Android 4.4.)_
1. Create a Classic Titanium app with the below "tiapp.xml" settings.
2. Build and run "CameraPhotoExternalTest.js" attached to [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059).
3. Tap on the "Take Picture" button.
4. On Android 6.0 and higher, verify app requests permission. Agree to it.
5. Verify a camera window appears.
6. Tap the Back button to go back to the app. Verify app does not exit.
7. Tap on the "Take Picture" button.
8. Take a picture and hit okay in the camera window.
9. Verify the photo you taken is displayed by the Titanium app.
10. In the JS code, uncomment line: `// saveToPhotoGallery: true`
11. Repeat steps 3-9.
12. Go to the Photos/Gallery app and verify you can find your photo there.
13. Build and run "CameraPhotoInternalTest.js" attached to [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059).
14. Repeat steps 3-12.

```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
			<uses-permission android:name="android.permission.CAMERA"/>
		</manifest>
	</android>
</ti:app>
```

---
**Camera Video Test:**
_(Run this test on Android 11 and Android 4.4.)_
1. Create a Classic Titanium app with the below "tiapp.xml" settings.
2. Build and run "CameraVideoExternalTest.js" attached to [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059).
3. Tap on the "Capture Video" button.
4. On Android 6.0 and higher, verify app requests permission. Agree to it.
5. Verify a camera window appears.
6. Tap the Back button to go back to the app. Verify app does not exit.
7. Tap on the "Capture Video" button.
8. Record video and hit okay in the camera window.
9. Verify the Titanium app displays and plays your video.
10. In the JS code, uncomment line: `// saveToPhotoGallery: true`
11. Repeat steps 3-9.
12. Go to the Photos/Gallery app and verify you can find your video there.
13. Build and run "CameraVideoInternalTest.js" attached to [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059).
14. Repeat steps 3-12.

```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
			<uses-permission android:name="android.permission.CAMERA"/>
			<uses-permission android:name="android.permission.RECORD_AUDIO"/>
		</manifest>
	</android>
</ti:app>
```

---
**Save Image File to Gallery Test:**
_(Run this test on Android 11 and Android 4.4.)_
1. Create a Classic Titanium app with the below "tiapp.xml" setting.
2. Build and run "ImageToGalleryTest.js" attached to [TIMOB-28059](https://jira.appcelerator.org/browse/TIMOB-28059).
3. Tap on the "Save Screenshot to Gallery" button.
4. Verify an alert displays it was successful.
5. Open the Photos/Gallery app and verify you can find the screenshot there.

```xml
<ti:app>
	<android>
		<manifest>
			<uses-sdk android:targetSdkVersion="30"/>
		</manifest>
	</android>
</ti:app>
```
